### PR TITLE
Remove $id from conditional schema except top level

### PR DESCRIPTION
**Pull request by tareq89**
Issue: https://github.com/opencrvs/opencrvs-core/issues/10350
### Problem
When defining conditionals using `defineConditional`, each schema was being assigned a `$id` based on the object hash of the schema itself.  
This caused issues when the same conditional schema was reused multiple times.  

For example:

```ts
conditionals: [
  {
    type: ConditionalType.SHOW,
    conditional: and(
      or(
        field('informant.relation').isEqualTo(InformantType.FATHER),
        field('mother.age').isEqualTo('40')
      ),
      or(
        and(
          not(field('father.ethnicity').isFalsy()),
          not(field('father.ethnicity').isEqualTo('None')),
          field('father.citizenship').isEqualTo('Filipino')
        ),
        not(field('father.citizenship').isEqualTo('Filipino'))
      )
    )
  }
]
```

Here, `field('father.citizenship').isEqualTo('Filipino')` is used twice.  
As a result, the same `$id` is generated twice in the final JSON Schema, which leads Ajv to throw the following error:

```
Error: reference "https://opencrvs.org/conditionals/4fb54005413214740f0a0bea138b7ea5eaf0ff57" 
resolves to more than one schema
```

### Solution
- Removed automatic `$id` assignment from every conditional schema.
- `$id` is now only added once at the **top-level schema**.  

This prevents Ajv from encountering duplicate `$id` values while still preserving reference integrity.

### Outcome
- No more duplicate `$id` conflicts in Ajv.  
- Safer reuse of conditional schemas across the form.  

## Checklist

- [ ] I have linked the correct Github issue under "Development"
- [ ] I have tested the changes locally, and written appropriate tests
- [ ] I have tested beyond the happy path (e.g. edge cases, failure paths)
- [ ] I have updated the changelog with this change (if applicable)
- [ ] I have updated the GitHub issue status accordingly

### CODE REVIEW

File: packages/commons/src/conditionals/conditionals.ts [hunk 1]
```diff
@@ -21,11 +21,30 @@ export type JSONSchema = {
...
+    for (const key of Object.keys(obj)) {
+      if (key === '$id') {
+        continue // skip $id
+      }
+      newObj[key] = removeIds(obj[key])
+    }
+    return newObj
+  }
+  return obj
+}
```
Comment (cibelius) [inline packages/commons/src/conditionals/conditionals.ts]: Below is a refactored version of the function, with early returns and no `continue`. Wonder if we could get rid of the `any`s here?

```suggestion
// eslint-disable-next-line @typescript-eslint/no-explicit-any
function removeIds(obj: any): any {
  if (Array.isArray(obj)) {
    return obj.map(removeIds)
  }

  if (obj === null || typeof obj !== 'object') {
    return obj
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const newObj: any = {}
  for (const key of Object.keys(obj)) {
    if (key !== '$id') {
      newObj[key] = removeIds(obj[key])
    }
  }
  return newObj
}
```

I did not test if this works, so please do make sure :)

Comment (cibelius) [reply packages/commons/src/conditionals/conditionals.ts]: Would be nice to have a comment for the function, telling what it does exactly and why its necessary.

Comment (tareq89) [reply packages/commons/src/conditionals/conditionals.ts]: Thanks for your refactor suggestion, I have adjusted the changes, please have a look.

**Review by cibelius — APPROVED** [review]

File: packages/commons/src/conditionals/conditionals.ts [hunk 2]
```diff
@@ -23,9 +24,14 @@ export type JSONSchema = {
...
 export function defineConditional(schema: any) {
+  // The same conditional schema may appear multiple times in the final schema.
+  // This causes duplicate $id values (since the hash is identical), and Ajv
+  // throws a `resolves to more than one schema` error.
+  // To avoid this, we only keep $id at the top level and remove it from all nested schemas.
+  const schemaWithooutIDRef = omitKeyDeep(schema, '$id')
   return {
-    $id: `https://opencrvs.org/conditionals/${objectHash.sha1(schema)}`,
-    ...schema
+    $id: `https://opencrvs.org/conditionals/${objectHash.sha1(schemaWithooutIDRef)}`,
```
Comment (rikukissa) [inline packages/commons/src/conditionals/conditionals.ts:33]: What if you would define $id in the `defineFormConditional` function below? That should always be the root definition. Make sure to also test that the events service memory usage doesn't start growing when you make this change and for instance reload the client app 100 times

**Review by Zangetsu101 — COMMENTED** [review]
I say define separate functions, one for top levels, where ids are generated and the internal recursive ones who call themselves

File: packages/commons/src/conditionals/conditionals.ts (hunk 2 above)
Comment (tareq89) [reply packages/commons/src/conditionals/conditionals.ts:33]: Correct me if I am wrong. We use only a compound conditional schema or single conditional schema, like:

```
and(
            or(
              field('informant.relation').isEqualTo(InformantType.FATHER),
              field('mother.age').isEqualTo('40')
            ),
            or(
              and(
                not(field('father.ethnicity').isFalsy()),
                not(field('father.ethnicity').isEqualTo('None')),
                field('father.citizenship').isEqualTo('Filipino')
              ),
              not(field('father.citizenship').isEqualTo('Filipino'))
            )
          )
```

or

```
field('father.dob').isBefore().now()
```

In both cases, `defineConditional` is the final definition of our schema.

For reference, in our current `commons/conditionals` file,

```

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function defineConditional(schema: any) {
  // The same conditional schema may appear multiple times in the final schema.
  // This causes duplicate $id values (since the hash is identical), and Ajv
  // throws a `resolves to more than one schema` error.
  // To avoid this, we only keep $id at the top level and remove it from all nested schemas.
  const schemaWithooutIDRef = omitKeyDeep(schema, '$id')
  return {
    $id: `https://opencrvs.org/conditionals/${objectHash.sha1(schemaWithooutIDRef)}`,
    ...schemaWithooutIDRef
  } as JSONSchema
}

export function defineFormConditional(schema: Record<string, unknown>) {
  const schemaWithForm = {
    type: 'object',
    properties: {
      $form: schema
    },
    required: ['$form']
  }

  return defineConditional(schemaWithForm)
}

```

We can see that `defineFormConditional` (which is called by all the `field('xxx').*` conditionals) ultimately calls `defineConditional`, as well as 'and(), or(), not()` etc also calls definedCondtionals.

So, `defineConditional` is the final function which is called, and we should decide here whether to remove or add $id.

Comment (tareq89) [reply packages/commons/src/conditionals/conditionals.ts:33]: Regarding the testing, I shall make sure of it.

**Review by rikukissa — APPROVED** [review]

**Comment by tareq89** [comment]
Ran bunch of e2es and track memory comsumption of events package. Doesn't look like there is a spike because of the changes in this PR:
```
109.9MB
109.9MB
109.9MB
109.9MB
109.9MB
109.9MB
109.9MB
109.9MB
109.9MB
109.9MB
109.9MB
109.9MB
109.9MB
109.9MB
109.9MB
109.9MB
109.9MB
109.9MB
99.7MB
99.0MB
99.0MB
96.7MB
84.2MB
123.4MB
132.9MB
140.3MB
140.9MB
140.9MB
140.9MB
140.9MB
140.9MB
140.9MB
140.9MB
140.9MB
140.9MB
140.9MB
140.9MB
140.9MB
140.9MB
140.9MB
140.9MB
140.9MB
140.9MB
140.9MB
140.9MB
140.9MB
140.9MB
140.9MB
149.7MB
158.1MB
157.5MB
173.4MB
187.2MB
187.2MB
187.2MB
205.0MB
224.6MB
224.6MB
263.8MB
269.7MB
299.1MB
299.1MB
299.1MB
299.1MB
297.9MB
297.9MB
297.9MB
297.9MB
297.9MB
297.9MB
297.9MB
299.3MB
299.3MB
299.3MB
299.3MB
299.3MB
299.3MB
299.3MB
299.3MB
299.3MB
299.3MB
299.3MB
299.3MB
299.3MB
299.6MB
302.3MB
304.9MB
306.3MB
308.1MB
282.6MB
271.0MB
271.0MB
271.0MB
271.0MB
261.9MB
259.3MB
259.3MB
258.6MB
255.6MB
255.6MB
246.2MB
245.0MB
245.0MB
247.3MB
251.4MB
253.3MB
253.3MB
253.3MB
257.3MB
257.3MB
257.3MB
257.3MB
257.3MB
257.2MB
257.2MB
257.2MB
257.2MB
257.2MB
257.2MB
255.9MB
255.9MB
255.9MB
255.9MB
255.9MB
255.9MB
255.9MB
255.9MB
138.5MB
114.8MB
114.8MB
114.8MB
114.8MB
114.8MB
114.8MB
114.8MB
114.8MB
114.8MB
114.8MB
114.8MB
114.4MB
111.7MB
111.7MB
111.7MB
111.7MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
111.6MB
139.2MB
145.9MB
145.9MB
145.9MB
145.9MB
152.4MB
152.4MB
157.8MB
170.6MB
184.6MB
189.8MB
189.8MB
189.8MB
189.8MB
201.6MB
216.0MB
226.2MB
235.1MB
252.1MB
274.6MB
277.2MB
285.7MB
291.0MB
293.7MB
293.7MB
294.3MB
301.8MB
301.8MB
301.8MB
301.8MB
301.8MB
301.8MB
313.7MB
322.8MB
333.2MB
330.7MB
325.9MB
330.6MB
330.6MB
331.2MB
334.9MB
338.4MB
340.6MB
340.6MB
340.6MB
340.6MB
340.6MB
340.6MB
340.6MB
340.6MB
340.6MB
340.6MB
340.6MB
345.0MB
348.4MB
348.4MB
333.7MB
333.7MB
333.7MB
336.7MB
337.1MB
337.1MB
337.7MB
340.6MB
340.6MB
340.6MB
341.2MB
341.2MB
341.2MB
341.2MB
341.2MB
344.1MB
348.5MB
331.5MB
331.4MB
337.6MB
340.4MB
340.3MB
340.3MB
340.3MB
340.3MB
341.5MB
340.3MB
340.9MB
342.9MB
333.3MB
333.4MB
331.6MB
331.6MB
330.2MB
336.1MB
339.8MB
343.4MB
343.4MB
347.6MB
347.6MB
348.1MB
348.1MB
348.1MB
348.1MB
348.1MB
348.1MB
348.1MB
348.1MB
348.1MB
348.1MB
334.8MB
334.8MB
333.2MB
333.2MB
333.1MB
333.1MB
330.1MB
330.8MB
330.2MB
330.2MB
332.7MB
332.6MB
332.6MB
332.6MB
332.6MB
338.9MB
340.6MB
340.7MB
346.3MB
351.7MB
351.7MB
351.7MB
351.7MB
351.7MB
351.7MB
351.7MB
352.4MB
351.7MB
351.7MB
351.7MB
351.7MB
351.7MB
351.1MB
351.1MB
351.1MB
351.1MB
351.2MB
353.1MB
353.1MB
353.1MB
353.1MB
353.1MB
353.1MB
353.1MB
358.9MB
364.2MB
367.2MB
371.5MB
353.8MB
353.8MB
353.8MB
353.8MB
353.8MB
353.8MB
356.2MB
356.3MB
356.4MB
356.4MB
357.7MB
356.5MB
356.5MB
356.5MB
356.5MB
356.5MB
356.5MB
356.7MB
358.4MB
358.4MB
358.4MB
358.4MB
358.4MB
358.4MB
358.4MB
359.0MB
359.0MB
361.5MB
364.1MB
352.5MB
352.5MB
354.9MB
354.9MB
358.1MB
356.9MB
357.1MB
357.4MB
360.4MB
365.3MB
351.1MB
351.1MB
351.3MB
351.3MB
351.3MB
351.3MB
351.3MB
351.3MB
351.3MB
353.4MB
352.3MB
352.4MB
352.4MB
352.4MB
352.4MB
352.4MB
352.4MB
354.3MB
354.3MB
354.3MB
354.3MB
354.3MB
354.3MB
354.2MB
349.7MB
349.2MB
326.6MB
293.6MB
293.7MB
169.3MB
169.3MB
169.3MB
90.2MB
85.0MB
85.0MB
85.0MB
129.3MB
147.7MB
150.9MB
157.0MB
168.0MB
173.1MB
174.2MB
174.2MB
191.8MB
195.9MB
194.8MB
189.4MB
186.1MB
167.3MB
166.9MB
166.9MB
166.9MB
159.8MB
167.9MB
193.1MB
201.4MB
202.8MB
202.8MB
202.8MB
202.8MB
204.8MB
224.3MB
226.9MB
245.0MB
240.2MB
237.2MB
232.7MB
232.7MB
232.7MB
232.7MB
236.4MB
240.8MB
239.6MB
246.3MB
285.3MB
303.6MB
301.8MB
301.7MB
298.5MB
285.9MB
285.9MB
283.7MB
283.7MB
283.9MB
293.0MB
295.9MB
296.0MB
293.8MB
293.8MB
296.9MB
289.7MB
291.9MB
294.7MB
272.6MB
272.6MB
200.1MB
120.3MB
102.7MB
100.2MB
99.5MB
142.4MB
168.8MB
168.8MB
168.8MB
168.8MB
168.8MB
168.9MB
213.4MB
216.8MB
233.4MB
236.8MB
229.4MB
227.6MB
226.5MB
224.8MB
224.5MB
198.6MB
198.4MB
197.3MB
197.3MB
197.3MB
197.3MB
197.3MB
197.3MB
197.3MB
219.2MB
241.4MB
242.3MB
244.7MB
244.7MB
244.7MB
244.7MB
244.7MB
244.7MB
244.7MB
244.7MB
244.7MB
244.7MB
244.7MB
236.8MB
238.9MB
238.9MB
239.1MB
240.7MB
241.9MB
244.2MB
244.2MB
244.2MB
253.1MB
256.6MB
289.8MB
289.8MB
310.9MB
310.9MB
300.3MB
300.3MB
300.3MB
300.3MB
300.3MB
300.3MB
300.3MB
300.3MB
300.3MB
300.3MB
300.3MB
294.7MB
294.6MB
294.1MB
226.9MB
143.7MB
103.9MB
103.0MB
42.3MB
40.7MB
36.9MB
36.8MB
36.8MB
36.8MB
86.7MB
86.6MB
86.6MB
86.6MB
86.6MB
92.4MB
143.9MB
157.3MB
160.3MB
196.1MB
206.0MB
223.4MB
254.8MB
267.6MB
251.5MB
253.8MB
256.6MB
257.8MB
255.8MB
255.8MB
255.8MB
255.8MB
255.8MB
258.2MB
258.9MB
258.9MB
270.6MB
280.4MB
280.4MB
280.4MB
280.4MB
280.4MB
280.4MB
280.4MB
299.0MB
299.1MB
319.0MB
335.5MB
335.5MB
349.9MB
349.9MB
341.0MB
339.4MB
342.9MB
342.9MB
344.9MB
345.5MB
345.4MB
345.4MB
345.4MB
332.1MB
301.0MB
301.0MB
301.0MB
301.0MB
301.0MB
245.4MB
194.1MB
159.0MB
159.0MB
159.0MB
159.0MB
108.4MB
92.9MB
92.6MB
97.8MB
147.1MB
165.8MB
165.8MB
165.9MB
168.8MB
171.9MB
196.1MB
200.7MB
221.7MB
240.8MB
244.6MB
271.9MB
275.9MB
297.3MB
297.4MB
297.4MB
297.4MB
297.4MB
297.4MB
293.6MB
289.5MB
294.5MB
296.4MB
296.7MB
296.8MB
299.0MB
311.8MB
301.4MB
300.2MB
304.3MB
305.6MB
309.7MB
309.7MB
309.7MB
309.7MB
309.6MB
306.5MB
285.7MB
285.5MB
239.0MB
239.2MB
228.9MB
223.9MB
223.6MB
223.1MB
222.0MB
222.0MB
222.0MB
222.0MB
222.0MB
237.5MB
251.3MB
253.6MB
265.2MB
258.6MB
258.9MB
275.9MB
296.9MB
305.7MB
305.7MB
305.7MB
305.7MB
305.7MB
305.7MB
305.7MB
305.7MB
305.7MB
305.7MB
305.7MB
306.6MB
308.5MB
313.7MB
313.7MB
313.9MB
317.2MB
317.3MB
318.1MB
330.5MB
330.6MB
330.3MB
333.2MB
337.0MB
323.4MB
333.3MB
341.7MB
344.5MB
347.0MB
321.6MB
324.2MB
324.3MB
324.3MB
318.9MB
317.7MB
317.7MB
317.3MB
300.3MB
300.3MB
299.9MB
301.1MB
301.7MB
301.7MB
301.7MB
302.4MB
302.4MB
302.5MB
302.6MB
307.0MB
307.6MB
308.0MB
315.0MB
327.8MB
328.4MB
333.2MB
313.2MB
313.2MB
313.2MB
313.2MB
313.2MB
313.2MB
313.2MB
313.2MB
313.2MB
313.2MB
317.8MB
317.8MB
317.9MB
321.4MB
321.4MB
321.4MB
321.4MB
322.6MB
328.0MB
330.6MB
319.3MB
325.1MB
324.5MB
331.9MB
341.9MB
347.7MB
354.5MB
328.8MB
329.2MB
331.1MB
331.4MB
335.0MB
335.8MB
335.8MB
335.8MB
335.8MB
335.8MB
335.8MB
335.8MB
335.8MB
335.8MB
335.8MB
335.8MB
335.8MB
324.7MB
309.0MB
193.5MB
125.2MB
125.2MB
122.6MB
121.0MB
120.1MB
120.1MB
120.1MB
120.1MB
120.1MB
120.1MB
122.6MB
175.7MB
187.9MB
217.8MB
245.9MB
265.1MB
268.8MB
292.7MB
296.3MB
295.5MB
295.5MB
295.5MB
295.5MB
295.5MB
295.5MB
295.5MB
295.5MB
295.5MB
295.5MB
295.5MB
295.5MB
314.7MB
315.5MB
315.5MB
315.5MB
315.5MB
315.5MB
315.5MB
313.7MB
313.7MB
313.2MB
313.6MB
318.8MB
321.7MB
323.0MB
328.6MB
318.1MB
325.2MB
311.7MB
311.7MB
306.3MB
306.3MB
303.6MB
303.7MB
304.0MB
304.1MB
306.1MB
306.1MB
302.7MB
302.7MB
302.7MB
302.7MB
302.7MB
302.7MB
302.7MB
302.7MB
302.7MB
301.5MB
300.0MB
300.0MB
300.0MB
300.0MB
300.0MB
300.0MB
300.0MB
300.0MB
300.0MB
301.1MB
301.9MB
301.9MB
```
